library/core/src/panic.rs

```diff
@@ -133,9 +147,27 @@ pub macro unreachable_2015 {
 #[allow_internal_unstable(panic_internals)]
 #[rustc_macro_transparency = "semitransparent"]
 pub macro unreachable_2021 {
-    () => (
-        $crate::panicking::panic("internal error: entered unreachable code")
-    ),
+    () => ({
+        #[cold]
+        #[track_caller]
+        #[inline(never)]
+        const fn panic_cold_explicit() -> ! {
+            $crate::panicking::panic("internal error: entered unreachable code");
+        }
+        panic_cold_explicit();
+    }),
+    // Special-case the single-argument case for const_panic.
+    ("{}", $arg:expr $(,)?) => ({
+        #[cold]
+        #[track_caller]
+        #[inline(never)]
+        #[rustc_const_panic_str] // enforce a &&str argument in const-check and hook this by const-eval
+        #[rustc_do_not_const_check] // hooked by const-eval
+        const fn panic_cold_display<T: $crate::fmt::Display>(arg: &T) -> ! {
+            $crate::panicking::panic($crate::format_args!("internal error: entered unreachable code: {}", *arg));
```

Comment:
Can this use panic_fmt and const_format_args?

unreachable_2015 and unreachable_2021 format string arms are also suspect, and the `($($t:tt)+)` for the one in 2021 is inconsistent. Not sure if that's a macro thing though.

Reply:
re:
> unreachable_2015 and unreachable_2021 format string arms are also suspect, and the ($($t:tt)+) for the one in 2021 is inconsistent. Not sure if that's a macro thing though.

I don't understand, what do you mean by this?

Reply:
Some of the cases for general `format_args!` usage use `($($t:tt)+)` instead of an `$fmt:expr` first. There's also some inconsistency with `panic!` being used within the macro itself. I think it can all be done with `panic_fmt`, `concat!`, and `const_format_args!`.

Reply:
Also, `unreachable_2015!` doesn't have a special case for "{}" for `const_panic`. These macros are very inconsistent.

Reply:
> These macros are very inconsistent

Agreed!